transport/http/headerlist.go

```diff
@@ -2,48 +2,120 @@ package http
 
 import (
 	"fmt"
+	"strconv"
 	"strings"
+	"unicode"
 )
 
 func splitHeaderListValues(vs []string, splitFn func(string) ([]string, error)) ([]string, error) {
-	for i := 0; i < len(vs); i++ {
-		if len(vs[i]) == 0 {
-			continue
-		}
+	values := make([]string, 0, len(vs))
 
+	for i := 0; i < len(vs); i++ {
 		parts, err := splitFn(vs[i])
 		if err != nil {
 			return nil, err
 		}
-		if len(parts) < 2 {
-			continue
-		}
-
-		tmp := make([]string, len(vs)+len(parts)-1)
-		copy(tmp, vs[:i])
-
-		for j, p := range parts {
-			tmp[i+j] = strings.TrimSpace(p)
-		}
-
-		copy(tmp[i+len(parts):], vs[i+1:])
-
-		vs = tmp
-		i += len(parts) - 1
+		values = append(values, parts...)
 	}
 
-	return vs, nil
+	return values, nil
 }
 
 // SplitHeaderListValues attempts to split the elements of the slice by commas,
 // and return a list of all values separated. Returns error if unable to
 // separate the values.
 func SplitHeaderListValues(vs []string) ([]string, error) {
-	return splitHeaderListValues(vs, commaSplit)
+	return splitHeaderListValues(vs, quotedCommaSplit)
 }
 
-func commaSplit(v string) ([]string, error) {
-	return strings.Split(v, ","), nil
+func quotedCommaSplit(v string) (parts []string, err error) {
+	v = strings.TrimSpace(v)
+
+	expectMore := true
+	for i := 0; i < len(v); i++ {
+		if unicode.IsSpace(rune(v[i])) {
```

Comment:
np: I think the cast to `rune` here is not required?  `v[i]` is already a rune when you index into a string.

Reply:
This is one of those odd differences between `v[i]` iteration vs `range`.  https://go.dev/play/p/ZPPf0pkQ8iQ 

`range` of string will return a `int` and `rune`(aka `int32`), but  a string index `v[i]`  will return `int` and `byte`(aka `uint8`)